drivers/rtc/rtc_ds1337.c

```diff
@@ -0,0 +1,747 @@
+	}
+
+	timeptr->tm_sec = bcd2bin(regs[0] & DS1337_SECONDS_MASK);
+	timeptr->tm_min = bcd2bin(regs[1] & DS1337_MINUTES_MASK);
+	timeptr->tm_hour = bcd2bin(regs[2] & DS1337_HOURS_MASK);
+	timeptr->tm_wday = bcd2bin(regs[3] & DS1337_DAY_MASK) + DS1337_DAY_OFFSET;
+	timeptr->tm_mday = bcd2bin(regs[4] & DS1337_DATE_MASK);
+	timeptr->tm_mon = bcd2bin(regs[5] & DS1337_MONTH_MASK) + DS1337_MONTH_OFFSET;
+	timeptr->tm_year = bcd2bin(regs[6] & DS1337_YEAR_MASK);
+	timeptr->tm_yday = -1;  /* Unsupported */
+	timeptr->tm_isdst = -1; /* Unsupported */
+	timeptr->tm_nsec = 0;   /* Unsupported */
+
+	/* Apply century offset */
+	if (regs[5] & DS1337_CENTURY_MASK) {
+		timeptr->tm_year += DS1337_TM_YEAR_2000;
+	}
+
+out_unlock:
+	ds1337_unlock_sem(dev);
+
+	if (!err) {
+		LOG_DBG("Read time: year: %d, month: %d, month day: %d, week day: %d, hour: %d, "
+			"minute: %d, second: %d",
+			timeptr->tm_year, timeptr->tm_mon, timeptr->tm_mday, timeptr->tm_wday,
+			timeptr->tm_hour, timeptr->tm_min, timeptr->tm_sec);
+	}
+
+	return err;
+}
+
+#ifdef CONFIG_RTC_ALARM
+
+static int ds1337_alarm_get_supported_fields(const struct device *dev, uint16_t id, uint16_t *mask)
+{
+	ARG_UNUSED(dev);
+
+	if (id == DS1337_ALARM_1_ID) {
+		*mask = DS1337_RTC_ALARM_TIME_1_MASK;
+		return 0;
+	} else if (id == DS1337_ALARM_2_ID) {
+		*mask = DS1337_RTC_ALARM_TIME_2_MASK;
+		return 0;
+	}
+
+	LOG_ERR("Invalid alarm ID: %d", id);
+	return -EINVAL;
+}
+
+static int ds1337_alarm_set_time(const struct device *dev, uint16_t id, uint16_t mask,
+				 const struct rtc_time *timeptr)
+{
+	const struct ds1337_config *config = dev->config;
+	uint8_t regs[4];
+	uint8_t reg_addr;
+	uint8_t reg_offset;
+	int err;
+
+	if (id >= DS1337_ALARMS_COUNT) {
+		LOG_ERR("Invalid alarm ID: %d", id);
+		return -EINVAL;
+	}
+
+	if ((mask & RTC_ALARM_TIME_MASK_MONTHDAY) && (mask & RTC_ALARM_TIME_MASK_WEEKDAY)) {
+		LOG_ERR("Month day and week day alarms cannot be set simultaneously");
+		return -EINVAL;
+	}
+
+	if (!ds1337_validate_alarm_mask(mask, id)) {
+		LOG_ERR("Unsupported mask 0x%04X for alarm %d", mask, id);
+		return -EINVAL;
+	}
+
+	if (!rtc_utils_validate_rtc_time(timeptr, mask)) {
+		LOG_ERR("Invalid alarm time");
+		return -EINVAL;
+	}
+
+	if (mask & RTC_ALARM_TIME_MASK_SECOND) {
+		regs[0] = bin2bcd(timeptr->tm_sec) & DS1337_ALARM_SECONDS_MASK;
+	} else {
+		regs[0] = DS1337_ALARM_DISABLE_MASK;
+	}
+
+	if (mask & RTC_ALARM_TIME_MASK_MINUTE) {
+		regs[1] = bin2bcd(timeptr->tm_min) & DS1337_ALARM_MINUTES_MASK;
+	} else {
+		regs[1] = DS1337_ALARM_DISABLE_MASK;
+	}
+
+	if (mask & RTC_ALARM_TIME_MASK_HOUR) {
+		regs[2] = bin2bcd(timeptr->tm_hour) & DS1337_ALARM_HOURS_MASK;
+	} else {
+		regs[2] = DS1337_ALARM_DISABLE_MASK;
+	}
+
+	if (mask & RTC_ALARM_TIME_MASK_MONTHDAY) {
+		regs[3] = bin2bcd(timeptr->tm_mday) & DS1337_ALARM_DATE_MASK;
+	} else if (mask & RTC_ALARM_TIME_MASK_WEEKDAY) {
+		regs[3] = bin2bcd(timeptr->tm_wday) & DS1337_ALARM_DAY_MASK;
+		regs[3] |= DS1337_DY_DT_MASK;
+	} else {
```

Comment:
need to apply `DS1337_DAY_OFFSET`?

```suggestion
	} else if (mask & RTC_ALARM_TIME_MASK_WEEKDAY) {
		regs[3] = bin2bcd(timeptr->tm_wday - DS1337_DAY_OFFSET) & DS1337_ALARM_DAY_MASK;
		regs[3] |= DS1337_DY_DT_MASK;
	} else {
```

Reply:
You're right! Fixed